File: retrocpu_test_framework_ts/src/mn1613/heap.ts
/**
 * テスト用ワードヒープ（ユーザ領域の malloc / free）
 * 根拠: test_framework.mdc / MN1613_CPUボードメモリ_IOマップ.mdc
 */

/** ユーザ領域先頭（ワード）。モニタ ROM/RAM・戻りスタブの直後 */
export const MN1613_USER_HEAP_START = 0x1800;

/** スタック領域先頭（ワード、排他）。ヒープはここより下 */
export const MN1613_USER_HEAP_END = 0xf800;

/** JsonTestSettings.malloc / セッションヒープ範囲 */
export type MallocSettings = {
  /** 先頭ワードアドレス。省略時 0x1800 */
  start?: number;
  /** 使用ワード数。省略時 start から 0xF800 まで */
  words?: number;
};

/**
 * 設定の malloc.start / malloc.words からヒープ範囲を求める。
 * @param malloc 設定。省略時はユーザ領域全体
 * @returns startWord（含む）と endWord（含まない）
 */
export function resolveMallocRange(malloc?: MallocSettings): {
  startWord: number;
  endWord: number;
  words: number;
} {
  const start = malloc?.start ?? MN1613_USER_HEAP_START;
  if (!Number.isInteger(start) || start < 0 || start > 0xffff) {
    throw new Error(`malloc.start: invalid word address ${start}`);
  }
  const words = malloc?.words ?? MN1613_USER_HEAP_END - start;
  if (!Number.isInteger(words) || words < 1) {
    throw new Error(`malloc.words: must be >= 1, got ${words}`);
  }
  const endWord = start + words;
  if (endWord > 0x10000) {
    throw new Error(
      `malloc: start 0x${hex4(start)} + words ${words} exceeds 16bit space`,
    );
  }
  return { startWord: start, endWord, words };
}

type HeapSpan = {
  start: number;
  size: number;
  used: boolean;
};

/**
 * 16bit ワードアドレスを 4 桁大文字 16 進にする。
 * @param n ワードアドレス
 * @returns 例 "1800"
 */
function hex4(n: number): string {
  return (n & 0xffff).toString(16).toUpperCase().padStart(4, "0");
}

/**
 * ワード単位の first-fit ヒープ。エミュメモリとは独立（アドレス帳のみ）。
 */
export class WordHeap {
  private readonly lo: number;
  private readonly hi: number;
  private spans: HeapSpan[] = [];

  /**
   * @param startWord 先頭ワード（含む）
   * @param endWord 末尾ワード（含まない）
   */
  constructor(
    startWord = MN1613_USER_HEAP_START,
    endWord = MN1613_USER_HEAP_END,
  ) {
    if (!Number.isInteger(startWord) || startWord < 0 || startWord > 0xffff) {
      throw new Error(`WordHeap: invalid start ${startWord}`);
    }
    if (
      !Number.isInteger(endWord) ||
      endWord <= startWord ||
      endWord > 0x10000
    ) {
      throw new Error(
        `WordHeap: end ${endWord} must be > start ${startWord} and <= 0x10000`,
      );
    }
    this.lo = startWord;
    this.hi = endWord;
    this.reset();
  }

  /** ヒープ先頭（含む） */
  get startWord(): number {
    return this.lo;
  }

  /** ヒープ末尾（含まない） */
  get endWord(): number {
    return this.hi;
  }

  /** 1 本の空き領域に戻す */
  reset(): void {
    this.spans = [{ start: this.lo, size: this.hi - this.lo, used: false }];
  }

  /**
   * 連続ワードを確保する。
   * @param wordCount ワード数（1 以上）
   * @returns 先頭ワードアドレス
   * @throws 0 以下、または空き不足
   */
  malloc(wordCount: number): number {
    if (!Number.isInteger(wordCount) || wordCount < 1) {
      throw new Error(`malloc: wordCount must be >= 1, got ${wordCount}`);
    }
    for (let i = 0; i < this.spans.length; i += 1) {
      const span = this.spans[i]!;
      if (span.used || span.size < wordCount) {
        continue;
      }
      if (span.size === wordCount) {
        span.used = true;
        return span.start;
      }
      this.spans.splice(
        i,
        1,
        { start: span.start, size: wordCount, used: true },
        { start: span.start + wordCount, size: span.size - wordCount, used: false },
      );
      return span.start;
    }
    throw new Error(
      `malloc: out of heap (${wordCount} words, 0x${hex4(this.lo)}–0x${hex4(this.hi)})`,
    );
  }

  /**
   * malloc で得た先頭アドレスを解放する。
   * @param wordAddr malloc の戻り値
   * @throws 未確保アドレス・二重解放
   */
  free(wordAddr: number): void {
    const addr = wordAddr & 0xffff;
    const i = this.spans.findIndex((s) => s.start === addr && s.used);
    if (i < 0) {
      throw new Error(`free: not an allocated block: 0x${hex4(addr)}`);
    }
    this.spans[i]!.used = false;
    this.coalesce();
  }

  /**
   * 確保中ブロックのワード数。
   * @param wordAddr malloc の戻り値
   * @returns ワード数
   */
  sizeOf(wordAddr: number): number {
    const addr = wordAddr & 0xffff;
    const span = this.spans.find((s) => s.start === addr && s.used);
    if (!span) {
      throw new Error(`sizeOf: not an allocated block: 0x${hex4(addr)}`);
    }
    return span.size;
  }

  /**
   * 確保中ブロックか。
   * @param wordAddr 先頭ワード
   * @returns 確保中なら true
   */
  isAllocated(wordAddr: number): boolean {
    const addr = wordAddr & 0xffff;
    return this.spans.some((s) => s.start === addr && s.used);
  }

  /** 隣接する空きを結合する */
  private coalesce(): void {
    const out: HeapSpan[] = [];
    for (const span of this.spans) {
      const last = out[out.length - 1];
      if (last && !last.used && !span.used) {
        last.size += span.size;
      } else {
        out.push({ start: span.start, size: span.size, used: span.used });
      }
    }
    this.spans = out;
  }
}

File: retrocpu_test_framework_rs/src/mn1613/heap.rs
use crate::error::FrameworkError;

pub const MN1613_USER_HEAP_START: u16 = 0x1800;
pub const MN1613_USER_HEAP_END: u16 = 0xf800;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct MallocSettings {
    pub start: Option<u16>,
    pub words: Option<u32>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MallocRange {
    pub start_word: u16,
    pub end_word: u32,
    pub words: u32,
}

pub fn resolve_malloc_range(malloc: Option<MallocSettings>) -> Result<MallocRange, FrameworkError> {
    let start = malloc
        .and_then(|m| m.start)
        .unwrap_or(MN1613_USER_HEAP_START) as u32;
    if start > 0xffff {
        return Err(FrameworkError::invalid_argument(format!(
            "malloc.start: invalid word address {start}"
        )));
    }

    let words = malloc
        .and_then(|m| m.words)
        .unwrap_or(MN1613_USER_HEAP_END as u32 - start);
    if words < 1 {
        return Err(FrameworkError::invalid_argument(format!(
            "malloc.words: must be >= 1, got {words}"
        )));
    }

    let end_word = start + words;
    if end_word > 0x1_0000 {
        return Err(FrameworkError::invalid_argument(format!(
            "malloc: start 0x{:04X} + words {} exceeds 16bit space",
            start, words
        )));
    }

    Ok(MallocRange {
        start_word: start as u16,
        end_word,
        words,
    })
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct HeapSpan {
    start: u16,
    size: u32,
    used: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WordHeap {
    lo: u16,
    hi: u32,
    spans: Vec<HeapSpan>,
}

fn hex4(n: u16) -> String {
    format!("{:04X}", n)
}

impl WordHeap {
    pub fn new(start_word: u16, end_word: u32) -> Result<Self, FrameworkError> {
        let start = start_word as u32;
        if end_word <= start || end_word > 0x1_0000 {
            return Err(FrameworkError::invalid_argument(format!(
                "WordHeap: end {end_word} must be > start {start} and <= 0x10000"
            )));
        }

        let mut out = Self {
            lo: start_word,
            hi: end_word,
            spans: Vec::new(),
        };
        out.reset();
        Ok(out)
    }

    pub fn with_default_range() -> Result<Self, FrameworkError> {
        Self::new(MN1613_USER_HEAP_START, MN1613_USER_HEAP_END as u32)
    }

    pub fn start_word(&self) -> u16 {
        self.lo
    }

    pub fn end_word(&self) -> u32 {
        self.hi
    }

    pub fn reset(&mut self) {
        self.spans.clear();
        self.spans.push(HeapSpan {
            start: self.lo,
            size: self.hi - self.lo as u32,
            used: false,
        });
    }

    pub fn malloc(&mut self, word_count: u32) -> Result<u16, FrameworkError> {
        if word_count < 1 {
            return Err(FrameworkError::invalid_argument(format!(
                "malloc: word_count must be >= 1, got {word_count}"
            )));
        }

        for i in 0..self.spans.len() {
            let span = self.spans[i];
            if span.used || span.size < word_count {
                continue;
            }

            if span.size == word_count {
                self.spans[i].used = true;
                return Ok(span.start);
            }

            self.spans.splice(
                i..=i,
                [
                    HeapSpan {
                        start: span.start,
                        size: word_count,
                        used: true,
                    },
                    HeapSpan {
                        start: span.start.wrapping_add(word_count as u16),
                        size: span.size - word_count,
                        used: false,
                    },
                ],
            );
            return Ok(span.start);
        }

        Err(FrameworkError::invalid_argument(format!(
            "malloc: out of heap ({} words, 0x{}-0x{:04X})",
            word_count,
            hex4(self.lo),
            self.hi
        )))
    }

    pub fn free(&mut self, word_addr: u16) -> Result<(), FrameworkError> {
        let pos = self
            .spans
            .iter()
            .position(|s| s.start == word_addr && s.used)
            .ok_or_else(|| {
                FrameworkError::invalid_argument(format!(
                    "free: not an allocated block: 0x{}",
                    hex4(word_addr)
                ))
            })?;

        self.spans[pos].used = false;
        self.coalesce();
        Ok(())
    }

    pub fn size_of(&self, word_addr: u16) -> Result<u32, FrameworkError> {
        self.spans
            .iter()
            .find(|s| s.start == word_addr && s.used)
            .map(|s| s.size)
            .ok_or_else(|| {
                FrameworkError::invalid_argument(format!(
                    "size_of: not an allocated block: 0x{}",
                    hex4(word_addr)
                ))
            })
    }

    pub fn is_allocated(&self, word_addr: u16) -> bool {
        self.spans.iter().any(|s| s.start == word_addr && s.used)
    }

    fn coalesce(&mut self) {
        let mut out: Vec<HeapSpan> = Vec::with_capacity(self.spans.len());
        for span in &self.spans {
            if let Some(last) = out.last_mut() {
                if !last.used && !span.used {
                    last.size += span.size;
                } else {
                    out.push(*span);
                }
            } else {
                out.push(*span);
            }
        }
        self.spans = out;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_range_defaults() {
        let r = resolve_malloc_range(None).expect("default range should work");
        assert_eq!(r.start_word, MN1613_USER_HEAP_START);
        assert_eq!(r.end_word, MN1613_USER_HEAP_END as u32);
    }

    #[test]
    fn heap_alloc_free_cycle() {
        let mut h = WordHeap::new(0x1800, 0x1810).expect("heap should be created");
        let a = h.malloc(4).expect("alloc A");
        let b = h.malloc(2).expect("alloc B");
        assert_eq!(a, 0x1800);
        assert_eq!(b, 0x1804);
        assert!(h.is_allocated(a));
        assert_eq!(h.size_of(b).expect("size of B"), 2);
        h.free(a).expect("free A");
        assert!(!h.is_allocated(a));
    }
}
